### ISSUE TITLE
Precision Improvement Suggestions from Qwen3 Coder

### ISSUE THREAD

**Issue by sempervictus** [guoqingbao/attention.rs#10]
Started looking into the gemma3 issues we're seeing in `vllm.rs` and figured i'd feed the diff off multigpu branch and sampling branch to qwen3 for a sense of how it thinks precision and accuracy can be improved. Here's what it thinks:

> #### **1. Current Implementation Overview**
> The hardware sampler implementation in `src/sampler.rs` and associated CUDA kernels (`src/kernels/src/gpu_sampling.cu`, `src/kernels/src/gpu_sampling.cuh`) uses a two-stage algorithm:
> - **Stage A**: Local top-K extraction per tile (using block radix sort).
> - **Stage B**: Global merging, softmax, top-P filtering, and sampling (using deterministic Philox RNG).
> 
> The implementation supports `f32`, `f16`, and `bf16` dtypes, with CUDA and Metal backends. Key features include:
> - Tile-based parallelism for large vocabularies.
> - Sliding window support via `sliding_window` parameter.
> - Deterministic sampling via `seed` and `token_pos` keys.
> 
> ---
> 
> #### **2. Identified Precision Gaps and Recommendations**
> 
> ##### **A. Floating-Point Precision in Softmax and Sampling**
> **Issue**:  
> The current softmax implementation uses `f32` accumulation internally, but the input logits may be `f16` or `bf16`. This introduces precision loss during softmax normalization, especially for large vocabularies where exponentiation amplifies rounding errors.
> 
> **Example**:  
> For `f16` logits with extreme values (e.g., `logits = [-1000.0, 1000.0]`), the softmax computation in `f32` may underflow to `0.0` for the negative value and overflow to `1.0` for the positive value, resulting in **zero probability for the correct token**.
> 
> **Improvement**:  
> Use **higher-precision accumulation** (`f64` or `f32` with fused operations) during softmax and top-P filtering.
> 
> **Implementation**:
> ```rust
> // In gpu_sampling.cuh, replace:
> // float max_val = ...;
> // float sum = ...;
> // with:
> double max_val = ...; // Use double for stability
> double sum = ...;
> ```
> 
> **Kernel Modification**:
> ```cuda
> // In stage B (merge and sample), replace:
> float probs[K];
> // with:
> double probs[K]; // Higher precision for probability accumulation
> ```
> 
> **Verification**:
> - Test with extreme logits (e.g., `logits = [1e4, -1e4]`).
> - Measure entropy loss vs. reference `f64` implementation.
> 
> ---
> 
> ##### **B. Top-P Filtering Precision**
> **Issue**:  
> Top-P filtering uses `f32` for cumulative probability checks. For small `p` (e.g., `p = 0.01`), rounding errors can cause **incorrect token inclusion/exclusion**.
> 
> **Example**:  
> If cumulative probability is `0.009999` due to rounding, the token is excluded even though it should be included (since `0.009999 < 0.01`).
> 
> **Improvement**:  
> Use **fixed-point arithmetic** or **compensated summation** for cumulative probability checks.
> 
> **Implementation**:
> ```cuda
> // In stage B, replace:
> float cum_prob = 0.0f;
> for (int i = 0; i < k; i++) {
>     cum_prob += probs[i];
>     if (cum_prob >= top_p) break;
> }
> // with:
> double cum_prob = 0.0;
> double compensation = 0.0;
> for (int i = 0; i < k; i++) {
>     double next_prob = probs[i] - compensation;
>     double new_cum = cum_prob + next_prob;
>     compensation = (new_cum - cum_prob) - next_prob;
>     cum_prob = new_cum;
>     if (cum_prob >= top_p) break;
> }
> ```
> 
> **Verification**:  
> - Test with `p = 0.01` and `k = 1000` tokens.
> - Ensure no token is excluded due to rounding.
> 
> ---
> 
> ##### **C. Random Number Generation (RNG) Precision**
> **Issue**:  
> The current implementation uses `Philox` RNG with `uint32_t` keys. For large vocabularies (`V > 2^32`), this causes **collisions** and non-uniform sampling.
> 
> **Example**:  
> If `V = 10^6`, the RNG key space (`2^32`) is sufficient, but for `V = 10^9`, the probability of collision increases, leading to **non-deterministic sampling**.
> 
> **Improvement**:  
> Use **64-bit RNG keys** (`uint64_t`) and **split the seed** into `seed` and `token_pos`.
> 
> **Implementation**:
> ```rust
> // In src/sampler.rs, update:
> fn next_token_pos(&self) -> u64 {
>     let current = self.token_pos.fetch_add(1, Ordering::Relaxed);
>     if current >= u32::MAX as u64 {
>         self.token_pos.store(0, Ordering::Relaxed);
>         0
>     } else {
>         current
>     }
> }
> 
> // In GPU kernel, use:
> uint64_t key = ((uint64_t)seed << 32) | token_pos;
> ```
> 
> **Verification**:  
> - Test with `V = 10^9` and `p = 0.9`.
> - Confirm uniformity via chi-squared test.
> 
> ---
> 
> ##### **D. Sliding Window Handling Precision**
> **Issue**:  
> The sliding window logic (`sliding_window` parameter) uses integer arithmetic for masking, which can cause **incorrect token masking** for non-aligned window sizes.
> 
> **Example**:  
> If `seq_len = 1025` and `sliding_window = 1024`, the mask should include tokens `1..1024`, but integer division may exclude token `1024`.
> 
> **Improvement**:  
> Use **floating-point arithmetic** for window bounds and **rounding** to avoid off-by-one errors.
> 
> **Implementation**:
> ```cuda
> // In paged_attention.cuh, replace:
> int start_token_idx = (seq_len - sliding_window) > 0 ? (seq_len - sliding_window) : 0;
> // with:
> int start_token_idx = __float2int_rd(seq_len - sliding_window);
> if (start_token_idx < 0) start_token_idx = 0;
> ```
> 
> **Verification**:  
> - Test with `seq_len = 1025`, `sliding_window = 1024`.
> - Confirm token `1024` is included.
> 
> ---
> 
> ##### **E. Memory Coalescing in Top-K Extraction**
> **Issue**:  
> The block radix sort in `stageA` uses uncoalesced memory access patterns, leading to **reduced GPU occupancy** and **lower throughput**.
> 
> **Example**:  
> For `V = 32768`, the sort kernel may issue 32768 non-coalesced reads/writes, reducing memory bandwidth utilization.
> 
> **Improvement**:  
> Use **coalesced memory access** via shared memory and **thread-block tiling**.
> 
> **Implementation**:
> ```cuda
> // In gpu_sampling.cuh, modify stageA:
> __shared__ float shared_probs[256]; // 256 threads per block
> __shared__ int shared_indices[256];
> 
> // Load coalesced data:
> int tid = threadIdx.x;
> int idx = blockIdx.x * blockDim.x + tid;
> if (idx < V) {
>     shared_probs[tid] = probs[idx];
>     shared_indices[tid] = idx;
> }
> __syncthreads();
> 
> // Perform radix sort on shared memory (coalesced).
> ```
> 
> **Verification**:  
> - Profile with `nvprof` to confirm memory bandwidth increase.
> - Measure latency reduction for `V = 32768`.
> 
> ---
> 
> #### **3. Summary of Critical Improvements**
> | **Area**                | **Current Issue**                     | **Improvement**                              | **Verification Metric**               |
> |--------------------------|----------------------------------------|-----------------------------------------------|----------------------------------------|
> | Softmax Precision        | `f32` accumulation for `f16`/`bf16`    | Use `f64` or fused `f32` for stability        | Entropy loss < 0.01 bit                |
> | Top-P Filtering          | Rounding errors in cumulative sum      | Compensated summation                         | Token inclusion error rate < 0.001%    |
> | RNG Keys                 | `uint32_t` limits for large vocabularies | Use `uint64_t` keys                           | Collision rate < 1e-9                  |
> | Sliding Window Masking   | Integer division off-by-one            | Use `__float2int_rd` for bounds               | Mask correctness at all window sizes   |
> | Memory Access            | Non-coalesced in top-K sort            | Shared memory tiling with coalesced loads     | Memory bandwidth > 80% of peak         |
> 
>

**Comment by guoqingbao** [comment]
Thanks, I think we switched to fp32 sampling and together with your patch for top_p in this repo, and latest updates of vllm.rs for tooling, it seems it can work well with claude code (at least for unquantized mirothinker 30B model based on my tests).

**Comment by guoqingbao** [comment]
Resolved.
